torch/csrc/jit/frontend/ir_emitter.cpp

```diff
@@ -3689,24 +3754,79 @@ struct to_ir {
-                                    const std::vector<Value*>& values,
-                                    TreeList trees) {
-          for (size_t i = 0, N = values.size(); i < N; ++i) {
-            std::stringstream ss;
-            if (!values[i]->type()->isSubtypeOfExt(type, &ss)) {
-              throw ErrorReport(trees[i])
-                  << "Dict " << what
-                  << " must contain only a single type, expected: "
-                  << type->repr_str() << " but found "
-                  << values[i]->type()->repr_str() << " instead.\n"
-                  << ss.str();
+        for (size_t i = 0, N = keys.size(); i < N; ++i) {
+          std::stringstream ss;
+          if (!keys[i]->type()->isSubtypeOfExt(key_type, &ss)) {
+            throw ErrorReport(key_trees[i])
+                << "Dict keys must contain "
+                << "only a single type, expected: " << key_type->repr_str()
+                << " but found " << keys[i]->type()->repr_str() << " instead.\n"
+                << ss.str();
+          }
+        }
+
+        if (!values.empty()) {
+          // Get the smallest common supertype of all the values. We
+          // need to use the O(N^2) algorithm instead of using
+          // `std::accumulate` to account for the case that two types
+          // cannot themselves be unified, but there exists some mutual
+          // parent type later in `values`. (E.g. T2 <: T1, T3 <: T1,
+          // `values` = {T2, T3, T1}. The real supertype should be `T1`,
+          // not `Any`)
+
+          // List of possible parent types
+          std::vector<TypePtr> seen;
+
+          auto get_greater_type =
+              [](const TypePtr t1, const TypePtr t2) -> c10::optional<TypePtr> {
+            if (t1->isSubtypeOf(t2)) {
+              return t2;
+            } else if (t2->isSubtypeOf(t1)) {
+              return t1;
+            } else {
+              return c10::nullopt;
+            }
+          };
+
+          auto value_types =
+              c10::fmap(values, [](Value* v) { return v->type(); });
+
+          for (const TypePtr type : value_types) {
+            bool added = false;
+            for (size_t i = 0; i < seen.size(); ++i) {
+              // Don't bother compare if it's already in our list of
+              // possible supertypes
+              if (type == seen[i]) {
+                continue;
+              }
+              auto maybe_parent = get_greater_type(type, seen[i]);
+              if (maybe_parent && *maybe_parent == type) {
+                seen[i] = type;
+                added = true;
+              }
+            }
+            if (!added) {
+              seen.push_back(type);
             }
           }
-        };
-        checkTypeOfValues(key_type, "keys", keys, key_trees);
-        checkTypeOfValues(value_type, "values", values, value_trees);
+
+          TypePtr inferred_type = seen.size() == 1 ? seen[0] : AnyType::get();
```

Comment:
Could you clarify how this condition works for the motivating use case on line 3773?

It looks like that, at end of algorithm, `seen` would be [T_1, T_1], which would fail check of `seen.size() == 1`, making the final type `Any`. If this is indeed the case, we can change type of `seen` to `std::unordered_set`, which fixes this issue by automatically deduping elements. Additionally, `unordered_set` has faster look up time, which makes it easier to check if a `type` has been visited before.

Reply:
Good point. I'm actually going to open a PR to change `unifyTypeList` to this more forward compatible version (anticipating a future in which we have a less flat type hierarchy)

Reply:
Could you clarify how "changing unityTypeList" can address the incorrect result for motivating use case?

Reply:
This algorithm is basically doing what `unifyTypeList` is doing; it just accounts for the subtyping relationships you pointed out. Because the number of types in our types system is pretty small, changing `unifyTypeList` from the current `O(N)` algorithm to a forward-compatible `O(N^2)` algorithm would bolster us against the future without causing any real performance degradation. If we introduce polymorphic types in the next year or so and our type system expands greatly, we would need the `O(N^2)` algorithm anyway. So why not go ahead and change it?

We should also make our list emission forwards compatible (which it currently isn't). The new `unifyTypeList` could be used in all four sections of `ir_emitter.cpp` that this PR touches.